dace/autodiff/data_forwarding/store.py

```diff
@@ -0,0 +1,683 @@
+        Given an edge leading an AccessNode or a map to the target node in the forward state,
+        add a path from the connector for this AccessNode to store its values for all iterations.
+        This can increase the dimension of the array. i.e. the size of the stored array is
+        greater or equal to the size of the original array.
+
+        :param edge: the edge connecting the AccessNode to save data from to a map node.
+        :return: the new AccessNode which contains the stored data,
+                 a list of memlets connecting an assign tasklet to this new AccessNode.
+        """
+
+    # Get the connector and edge to save
+    if isinstance(edge.src, nodes.AccessNode) and edge.src is not forward_an:
+
+        # Get the incoming edge to this AccessNode
+        in_edges = forward_state.in_edges(edge.src)
+
+        # There should only be one incoming edge
+        assert len(in_edges) == 1
+
+        # Get the memlet path for the edge incoming to this AccessNode
+        memlet_path = forward_state.memlet_path(in_edges[0])
+
+        # The start of this path should be the forward AccessNode
+        assert forward_an is memlet_path[0].src
+
+        # The last edge in the memlet path has the connector we want to save
+        edge = memlet_path[-1]
+
+    # Add a new AccessNode and array to the forward pass
+    # First, check if a stored array with this name already exists
+    new_store_node_name = forward_state.sdfg._find_new_name("stored_" + forward_an.data)
+
+    # Get the new array shape
+    # This will be the shape of the current array
+    shape: List[int] = list(bwd_generator.sdfg.arrays[forward_an.data].shape)
+
+    # If the shape is an expression:
+    if any((isinstance(s, dace.symbol) or isinstance(s, sp.Expr)) and not (str(s) in bwd_generator.sdfg.free_symbols)
+           for s in shape):
+        # Otherwise, replace all the loop dependant allocations with the max length of the loop
+        # For example, an array of size [i+1] in a range(2, 10) loop will be stored in a [10, 10] array (1)
+        # Additionally, an array of size [32-i] in the same loop will be stored in a [10, 30]  (2)
+        loops = _get_all_enclosing_loops(forward_state)
+
+        if len(loops) > 0:
+            # Loop over the shape dimensions
+            for i, s in enumerate(shape):
+                if ad_utils.shape_has_symbols_to_replace(bwd_generator.sdfg, s):
+                    loop_size, loop_index = _get_symbol_upper_bound_from_loop(bwd_generator, s, loops)
+                    # Replace the symbol with the loop size and evaluate the expression
+                    # Check if loop size can be converted to an integer
+                    if isinstance(loop_size, int) or (isinstance(loop_size, str) and ad_utils.is_int(loop_size)):
+                        shape[i] = s.subs(sp.Symbol(loop_index), loop_size)
+                    else:
+                        shape[i] = s.subs(sp.Symbol(loop_index), dace.symbol(loop_size))
```

Comment:
dangerous I think (because of properties of symbols, e.g., nonnegative). Maybe use `symbolic.pystr_to_symbolic(loop_index)` and size instead?